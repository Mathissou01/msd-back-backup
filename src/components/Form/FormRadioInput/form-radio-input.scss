@use "src/styles/01-settings/settings.colors" as colors;

.c-FormRadioInput {
  display: flex;
  flex-direction: column;

  &__Options {
    display: flex;
    gap: 20px;

    &_vertical {
      flex-direction: column;
    }

    &_horizontal {
      flex-flow: row wrap;
    }
  }

  &__Option {
    display: flex;
    flex-direction: row;
    gap: 8px;
    align-items: center;
  }

  &__Input {
    position: relative;
    width: 20px;
    height: 20px;
    color: colors.$expert-blue;
    background-color: transparent;
    border: 2px solid colors.$expert-blue;
    border-radius: 50%;
    cursor: pointer;
    appearance: none;

    &_checked::after {
      position: absolute;
      top: 3px;
      right: 3px;
      bottom: 3px;
      left: 3px;
      background-color: colors.$expert-blue;
      border-radius: 50%;
      content: "";
    }
  }
}
