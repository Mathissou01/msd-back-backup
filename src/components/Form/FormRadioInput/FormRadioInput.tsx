import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import React, { useEffect } from "react";
import FormLabel from "../FormLabel/FormLabel";
import "./form-radio-input.scss";

interface IOption {
  value: string | number;
  label: string;
}

interface IFormRadioInputProps {
  name: string;
  displayName: string;
  secondaryDisplayName?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  options: Array<IOption>;
  defaultValue?: string;
  onChange?: (data: unknown) => void;
}

export default function FormRadioInput({
  name,
  displayName,
  secondaryDisplayName,
  isRequired = false,
  isDisabled = false,
  options,
  defaultValue,
  onChange,
}: IFormRadioInputProps) {
  /* Static Data */
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const {
    register,
    setValue,
    watch,
    formState: { isSubmitting, errors },
  } = useFormContext();
  const watchChecked: string | number = watch(name);

  useEffect(() => {
    if (watchChecked === undefined && defaultValue) {
      setValue(name, defaultValue);
    }
  }, [defaultValue, name, setValue, watchChecked]);

  useEffect(() => {
    if (watchChecked && onChange) {
      onChange(watchChecked);
    }
    // eslint-disable-next-line
  }, [watchChecked]);

  return (
    <div className="c-FormRadioInput">
      <fieldset>
        <FormLabel
          label={displayName}
          isRequired={isRequired}
          secondaryLabel={secondaryDisplayName}
          tagType="legend"
        />
        <div className="c-FormRadioInput__Options">
          {options.map((option, index) => (
            <div key={name + index} className="c-FormRadioInput__Option">
              <input
                className={`c-FormRadioInput__Input ${
                  option.value.toString() === watchChecked
                    ? "c-FormRadioInput__Input_checked"
                    : ""
                }`}
                {...register(name, {
                  required: {
                    value: isRequired,
                    message: errorMessages.required,
                  },
                })}
                type="radio"
                id={name + index}
                value={option.value}
                checked={option.value.toString() === watchChecked}
                disabled={isSubmitting || isDisabled}
                data-testid={`form-radio-input_${index}`}
              />
              <FormLabel forId={name + index} label={option.label} />
            </div>
          ))}
        </div>
      </fieldset>
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <p>{message ?? "Error"}</p>
        )}
      />
    </div>
  );
}
