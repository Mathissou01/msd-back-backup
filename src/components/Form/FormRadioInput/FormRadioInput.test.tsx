import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormRadioInput from "./FormRadioInput";

const mock = {
  name: "form-radio-input",
  displayName: "form radio input",
  secondaryDisplayName: "secondary display name",
  secondaryLabel: "secondary label",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormRadioInput", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <FormRadioInput
          name={mock.name}
          displayName={mock.displayName}
          secondaryDisplayName={mock.secondaryDisplayName}
          isRequired={false}
          isDisabled={false}
          options={[
            { value: 0, label: "first" },
            { value: 1, label: "second" },
          ]}
          defaultValue={"1"}
        />
      </Wrapper>,
    );
    expect(screen.getByTestId("form-radio-input_0")).toBeInTheDocument();
    expect(screen.getByTestId("form-radio-input_1")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
