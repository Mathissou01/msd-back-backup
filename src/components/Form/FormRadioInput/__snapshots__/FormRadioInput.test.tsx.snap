// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRadioInput renders 1`] = `
<div>
  <div
    class="c-FormRadioInput"
  >
    <fieldset>
      <div
        class="c-FormLabel"
      >
        <legend
          class="c-FormLabel__LabelWrapper"
        >
          <span
            class="c-FormLabel__Label"
          >
            form radio input
          </span>
        </legend>
        <legend
          class="c-FormLabel__LabelWrapper"
        >
          <span
            class="c-FormLabel__Secondary"
          >
            secondary display name
          </span>
        </legend>
      </div>
      <div
        class="c-FormRadioInput__Options"
      >
        <div
          class="c-FormRadioInput__Option"
        >
          <input
            class="c-FormRadioInput__Input "
            data-testid="form-radio-input_0"
            id="form-radio-input0"
            name="form-radio-input"
            type="radio"
            value="0"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-radio-input0"
            >
              <span
                class="c-FormLabel__Label"
              >
                first
              </span>
            </label>
          </div>
        </div>
        <div
          class="c-FormRadioInput__Option"
        >
          <input
            class="c-FormRadioInput__Input c-FormRadioInput__Input_checked"
            data-testid="form-radio-input_1"
            id="form-radio-input1"
            name="form-radio-input"
            type="radio"
            value="1"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-radio-input1"
            >
              <span
                class="c-FormLabel__Label"
              >
                second
              </span>
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;
